### DISCUSSION TITLE
Cast exception, prolly something silly, learning from zero

### DISCUSSION THREAD

**Discussion by djeikyb** [AvaloniaUI/Avalonia#8560] [Q&A] [answered]
I used the avalonia.mvvm template to make an app, but when I add a simple view, I get an exception.

I made a user control view and am trying to add it to the main window. My view contains a text block:

```xml
<UserControl xmlns="https://github.com/avaloniaui"
             xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
             xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
             xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
             mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
             xmlns:vm="using:Failure.ViewModels"
             x:DataType="vm:WhateverViewModel"
             x:CompileBindings="True"
             x:Class="Failure.Views.WhateverView">
    <TextBlock Text="{Binding SomeTextProperty}" />
</UserControl>
```

My view model is that single string property:

```cs
public class WhateverViewModel : ViewModelBase
{
    public string SomeTextProperty { get; } = "why doesn't this text show up, and why is there an exception";
}
```

And I added it to the main window like:

```xml
<Window xmlns="https://github.com/avaloniaui"
        xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
        xmlns:vm="using:Failure.ViewModels"
        xmlns:d="http://schemas.microsoft.com/expression/blend/2008"
        xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006"
        xmlns:views="clr-namespace:Failure.Views"
        mc:Ignorable="d" d:DesignWidth="800" d:DesignHeight="450"
        x:Class="Failure.Views.MainWindow"
        Icon="/Assets/avalonia-logo.ico"
        Title="Failure">
    <Design.DataContext>
        <vm:MainWindowViewModel/>
    </Design.DataContext>
    <StackPanel>
        <TextBlock Text="{Binding Greeting}" HorizontalAlignment="Center" VerticalAlignment="Center"/>
        <views:WhateverView />
    </StackPanel>
</Window>
```

When I run the app, my text doesn't show up. When I debug the app, there's this exception:

> System.InvalidCastException: Unable to cast object of type 'Failure.ViewModels.MainWindowViewModel' to type 'Failure.ViewModels.WhateverViewModel'.
>    at CompiledAvaloniaXaml.XamlIlHelpers.Failure.ViewModels.WhateverViewModel,Failure.SomeTextProperty!Getter(Object )
>    at Avalonia.Data.Core.ClrPropertyInfo.Get(Object target) in /_/src/Avalonia.Base/Data/Core/ClrPropertyInfo.cs:line 27
>    at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.get_Value() in /_/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/PropertyInfoAccessorFactory.cs:line 95
>    at Avalonia.Markup.Xaml.MarkupExtensions.CompiledBindings.InpcPropertyAccessor.SendCurrentValue() in /_/src/Markup/Avalonia.Markup.Xaml/MarkupExtensions/CompiledBindings/PropertyInfoAccessorFactory.cs:line 139

I've gone through the ToDo List and Music App tutorials. I'm mostly a web dev (vuejs experience), so this is all completely new to me. I added a [repo for the whole project][0], in case that helps.

Why is it trying to cast? _What_ is trying to cast? I'm still building a mental model of how all these pieces fit together.

[0]: https://github.com/djeikyb/AvaloniaHelp

**Comment by jp2masa** [comment, marked as answer]
The problem is that you defined `x:DataType="vm:WhateverViewModel"` on the user control, but the instance is created with `<views:WhateverView />`, which means the data context is "inherited", i.e. it will be the one from the closest parent with a data context, in this case the `Window`/`MainWindow`. That data context is a `MainViewModel`, and the user control expects a `WhateverViewModel`, so the cast fails.

Assuming there's a hierarchy where `MainViewModel` is the parent and `WhateverViewModel` is the child, you can do something like:

```cs
...
    public class MainWindowViewModel : ViewModelBase
    {
        public string Greeting => "Welcome to Avalonia!";

        public WhateverViewModel Whatever { get; } = new WhateverViewModel();
    }
```

```xaml
...
        <views:WhateverView DataContext="{Binding Whatever}" />
...
```

**Reply by YariLAN** [reply]
Thanks!!!!
